openmdao/core/group.py

```diff
@@ -1267,20 +1268,22 @@ def _solve_nonlinear(self):
 
         name = self.pathname if self.pathname else 'root'
 
-        # Execute guess_nonlinear if specified.
-        # We need to call this early enough so that any solver that needs initial guesses has
-        # them.
-        # TODO: It is pointless to run this ahead of non-iterative solvers.
-        for sub in self.system_iter(recurse=True):
-            if hasattr(sub, 'guess_nonlinear'):
-                with sub._unscaled_context(outputs=[sub._outputs], residuals=[sub._residuals]):
-                    sub.guess_nonlinear(sub._inputs, sub._outputs, sub._residuals)
-
         with Recording(name + '._solve_nonlinear', self.iter_count, self):
             result = self._nonlinear_solver.solve()
 
         return result
 
+    def _guess_nonlinear(self):
+        """
+        Provide initial guess for states.
+        """
+        from openmdao.api import ImplicitComponent
+        for isub, sub in enumerate(self._subsystems_myproc):
+            if overrides_method('guess_nonlinear', sub, ImplicitComponent) or \
```

Comment:
this check has some overhead, so I'm not sure we want to be doing it all of the time.  I would suggest adding a boolean flag to System, maybe _has_guess or something, default it to false, and in one of the setup methods in ImplicitComponent, do the check and set the flag.  That way we won't have to do the import of ImplicitComponent here either.  I guess in Group you might want to set the flag to True, although it wouldn't be hard to determine if any of the Group's children had the flag set to True and if not, leave it False.